driven/generic/adapter.py

```diff
@@ -175,18 +181,250 @@ def annotate_new_metabolite(metabolite):
+        """
+        logger.info('Add gene: {}'.format(feature.name))
+        identifier = feature_id(feature)
+        if self.model.genes.query(identifier, attribute='name'):  # do not add if gene is already there
+            logger.info('Gene {} exists in the model'.format(feature.name))
+            return
+        for reaction_id, equation in self.genes_to_reactions.get(identifier, {}).items():
+            self.add_reaction(reaction_id, equation, identifier)
+        logger.info('Gene added: {}'.format(identifier))
+
+    def add_reaction(self, reaction_id, equation, gene_name):
+        """Add new reaction by rn ID from equation, where metabolites defined by kegg ids.
+
+        :param reaction_id: reaction rn ID
+        :param equation: equation string, where metabolites are defined by kegg ids
+        :param gene_name: gene name
+        :return:
+        """
+        reaction = Reaction(reaction_id)
+        self.model.add_reactions([reaction])
+        equation = map_equation_to_bigg(equation, self.compartment)
+        logger.info('New reaction: {}'.format(equation))
+        reaction.build_reaction_from_string(equation)
+        for metabolite in reaction.metabolites:
+            if metabolite.formula is None:  # unknown metabolite
+                self.annotate_new_metabolite(metabolite)
+                self.create_exchange(metabolite)
+                self.changes['added']['metabolites'].add(metabolite)
+        reaction.gene_reaction_rule = gene_name
+        self.changes['added']['reactions'].add(reaction)
+
+
+class MediumChangeModel(ModelModificationMixin):
+    """
+    Applies medium on cameo model
+    """
+
+    def __init__(self, model, medium):
+        """
+        Parameters
+        ----------
+        model
+            cameo model
+        medium
+            list of dictionaries of format
+            {'id': <compound id (<database>:<id>, f.e. chebi:12345)>, 'concentration': <compound concentration (float)>}
+        """
+        self.medium = medium
+        self.model = model
+        self.changes = {
+            'added': {'reactions': set()},
+        }
+        self.apply_medium()
+
+    def apply_medium(self):
+        """For each metabolite in medium try to find corresponding metabolite in e compartment of the model.
+        If metabolite is found, change the lower limit of the reaction to a negative number,
+        so the model would be able to consume this compound.
+        If metabolite is not found in e compartment, log and continue.
+        """
+        for compound in self.medium:
+            existing_metabolite = self.model_metabolite(compound['id'], '_e')
+            if not existing_metabolite:
+                logger.info('No metabolite {}'.format(compound['id']))
+                continue
+            logger.info('Found metabolite {}'.format(compound['id']))
+            self.add_demand_reaction(existing_metabolite)
+
+
+class MeasurementChangeModel(ModelModificationMixin):
```

Comment:
See https://github.com/biosustain/driven/blob/master/driven/flux_analysis/fluxomics.py and https://github.com/biosustain/driven/blob/master/driven/data_sets/fluxes.py

Flux measurements are a data type like expression or concentrations.